### ISSUE TITLE
Add KeepKey hardware wallet support

### ISSUE THREAD

**Issue by flyswatter** [MetaMask/metamask-plugin#889]
A new contender in the ethereum hardware wallet space:

http://ethnews.com/keepkey-hard-wallet-supports-ethereum

Blocked by #328

**Comment by SCBuergel** [comment]
#328 is closed and references https://github.com/MetaMask/metamask-plugin/milestone/14 which is also closed. Does that mean this issue is basically ready to progress?

**Comment by flyswatter** [comment]
This is now blocked by UI work that we're doing, for how we're going to fit extra account types into MetaMask.

**Comment by keepkeyjon** [comment]
Is this still blocked?

We've published https://github.com/keepkey/keepkey.js, which should make it reasonably straightforward. I'm also happy to provide a couple of development devices to help with integration / testing, if you're interested.

**Comment by danfinlay** [comment]
Feel free to email me dan [at] metamask.io to continue the discussion. Some development devices would help, but you could also make some headway by following the lead of some of our other hardware wallet integrations.

https://github.com/MetaMask/metamask-extension/pulls?q=is%3Apr+Trezor+is%3Aclosed
https://github.com/MetaMask/metamask-extension/pulls?utf8=%E2%9C%93&q=is%3Apr+Ledger+is%3Aclosed+

Right now this is a bit lower priority than our mobile client, so it might be another quarter or two before we have the bandwidth to fully pursue this, but you could help us line up the effort.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.3 ETH (62.2 USD @ $207.33/ETH)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/MetaMask/metamask-extension/889/2982).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/MetaMask/metamask-extension/889/2982).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $95,179.15 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 5 months, 4 weeks from now.
Please review their action plans below:


**1) [abrahamh08](https://gitcoin.co/profile/abrahamh08) has started work.**

I can try using the KeepKey emulator to develop this

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/MetaMask/metamask-extension/889/2982).

**Comment by frankchen07** [comment]
@abrahamh08 - frank from gitcoin here - any progress on this task, or should we release it to the general public?

**Comment by keepkeyjon** [comment]
"Fun" thing I learned over the weekend: it turns out that the `chrome.hid` and `chrome.usb` APIs are only available to Chrome Apps, but not Chrome Extensions. That means you can't talk to a KeepKey from the background page in Metamask, since you need access to `window.navigator.usb` to make it work.

**Comment by keepkeyjon** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=352257#c7

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 4 months, 2 weeks from now.
Please review their action plans below:


**1) [deepsi43](https://gitcoin.co/profile/deepsi43) has started work.**

Hi,
I am interested in working on this project .

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/MetaMask/metamask-extension/889/2982).

**Comment by deepsi43** [comment]
Is this issue still open ??

**Comment by danfinlay** [comment]
It appears so!

**Comment by deepsi43** [comment]
I have been looking at the trezor integration but couldnt find any trace of keepkey modules added !!
Did check with the keepkey installation process and modules.

**Comment by keepkeyjon** [comment]
Any KeepKey integration with Metamask will be a wildly different beast than the Trezor integration, since we don't integrate with TrezorConnect. Don't expect it to be a simple copy-paste level of effort.

**Comment by Abrahamh08** [comment]
@frankchen Sorry I didn't see you mention me earlier, but no I decided that I didn't have time to work on it, sorry to put a hold on it

**Comment by roncoejr** [comment]
Good to see that there is an open feature request on this one.  Understand that the mobile client is taking priority.  Looking forward to seeing Keepkey support in MetaMask sometime in September.

All the best to the team.

**Comment by step21** [comment]
It appears this is still blocked by chrome not allowing use of usb by extensions, right?

**Comment by keepkeyjon** [comment]
I believe extensions have access to `window.navigator.usb`, but not `chrome.usb`, which means device communication has to happen from the popup, and not the background page.

**Comment by revonoc** [comment]
Any updates? Would love this integration but sadly don’t have the skills to help out

**Comment by frankchen07** [comment]
@deepsi43, frank from gitcoin here, it looks like you've started work. Do you have updated results? if not, I'll release this bounty back to the public.

**Comment by deepsi43** [comment]
Will get the update by the today evening@Frank

**Comment by revonoc** [comment]
Good to hear this is still being worked on!

**Comment by revonoc** [comment]
Bumping this again to see if it's still open - would love the Keepkey support for sure, but not sure how hard it is to integrate!

**Comment by danfinlay** [comment]
I have a feeling if you wrote a module like [eth-trezor-keyring](https://github.com/MetaMask/eth-trezor-keyring) for KeepKey, you could model it into a plugin for MetaMask using our [newly released plugin branch beta](https://github.com/MetaMask/metamask-snaps-beta/wiki).

You'd need to use [this branch](https://github.com/MetaMask/metamask-snaps-beta/pull/63) until it's merged, but this could be the model for future hardware wallet integrations on MetaMask to be more easily contributed.

**Comment by deepsi43** [comment]
I have been looking into the same aspect and have been working on it

**Comment by keepkeyjon** [comment]
FYI, we've replaced keepkey.js with the more general https://github.com/shapeshift/hdwallet

**Comment by deepsi43** [comment]
Yes I did have seen it.just wondering which would be a better approach like using keepkeyjs directly or to work based on the new hdwallet shapeshitftoss webusbadapter

**Comment by keepkeyjon** [comment]
keepkey.js is a dead project. I wouldn't recommend building new things on it.

**Comment by keepkeyjon** [comment]
I highly recommend building this on top of `@shapeshiftoss/hdwallet-keepkey-webusb` and the other related packages.

**Comment by revonoc** [comment]
@deepsi43 so glad you're working on it! Thanks man

**Comment by deepsi43** [comment]
Seems like I cannot continue to work on this module anymore!!!...

**Comment by walidmujahid** [comment]
@deepsi43 That is sad to hear. I hope everything is alright. Out of curiosity, what have accomplished so far? Where can I see the code that you have?

**Comment by lazaridiscom** [comment]
if this keepkey hw-wallet is similar (api/functionality) to trezor/ledger, then the easiest (dev) way would be to produce a 100% compatible eth-trezor-keyring (or ledger) as a drop-in replacement. Thus no need to change the UI of MM initially.

**Comment by keepkeyjon** [comment]
You'll need some small MetaMask UI tweaks, since PIN/Passphrase on Trezor are handled through TrezorConnect UI, which is not provided by @shapeshiftoss/hdwallet-keepkey nor keepkey.js.

**Comment by lazaridiscom** [comment]
The "fake" `eth-trezor-keyring` could provide this initially by mock-functionality (even with a hardcoded PIN/Passphrase).

**Comment by lazaridiscom** [comment]
See https://github.com/MetaMask/metamask-extension/issues/4060 for more relate info re ledger-implementation (though outdated, as today, webusb would be the way to go).

My estimation for KeepKey is 3 to 6 weeks for production-grade.

**Comment by walidmujahid** [comment]
If webusb is the way to go, perhaps it would be more worthwhile to re-implement things using packages from shapeshift/hdwallet?

**Comment by keepkeyjon** [comment]
@walidmujahid yes

**Comment by walidmujahid** [comment]
I thought so. I am not very comfortable with the quick dev solution of making a "fake" keyring and hardcoding the pin. 

I am still trying find some options that could combine a quick build and benefit MetaMask in the longterm, so I will comment more on this thread and or hopefully more discussion can illuminate on different but actionable solution paths. 

I think it is worth discussing whether avoiding design changes would be counterproductive in both the short term and long term, or just the long term. 

I feel that by shifting to building on shapeshift/hadwallet packages, that there could be value that could be added to design discussions to #731 which in turn can not only help with adding KeepKey wallet support to MM but perhaps also enable various features in regards to hardware wallets in MetaMask.

**Comment by walidmujahid** [comment]
I thought so. I am not very comfortable with the quick dev solution of making a "fake" keyring and hardcoding the pin.

Perhaps more discussion can illuminate on different but still actionable solution paths.

I think it is worth discussing whether avoiding design changes would be counterproductive in both the short term and long term, or just the long term.

It might be worth discussing whether shifting to building on shapeshift/hadwallet packages could lead to value that could be added to design discussions at #731 and help in part in achieving the idea of multiple keychains.

**Comment by lazaridiscom** [comment]
>>If webusb is the way to go, would it be more worthwhile to re-implement things using packages from shapeshift/hdwallet?
> @walidmujahid yes

@walidmujahid , @keepkeyjon , you are both wrong.

"Getting things done" on a code-base full off technical debt and in-development (metamask) needs some "higher grade abstract processing", which usually cannot be discussed much, as you cannot teach decades-xp and ability to discussion partners.

This task here is "peanuts", the "fake" implementation becomes quickly a "real" one. Doing "the right implementation" on a code-base like MetaMask, is not always possible. You need to adapt, and many times, "doing a bad implementation, compatible to existent code" is the logical way to go.

As for issue like #731 and the likes (there are many of them), this need a bottom-up redesign of MM, to eliminate technical-debt and create a high-quality code-base, well, this cannot be do as side-projects based on $50 bounties (which are not even worth the metal overhead they introduce).

This redesign was attempted during #4060, and is again attempted here #7601 (with a different starting point).

In fact "achieving the idea of multiple keychains." would not even exist for me. It would be a "natural by-product" of an elegant implementation (or a "higher-grade-redesign").

Enough talk, me out here

**Comment by lazaridiscom** [comment]
semi-related regression: https://github.com/MetaMask/metamask-extension/issues/3214#issuecomment-565250560

**Comment by walidmujahid** [comment]
I assume that the @shapeshiftoss/hdwallet-keepkey-webusb has the potential to build cross browser support unlike the hdwallet-keepkey-chromeusb package. 

I personally am convinced by @keepkeyjon about going with using shapeshift hadwallet-keepkey-{*} packages. That and I had personally read through all their code, so I feel a bit biased towards them.

Regarding design needs, I assume the main and perhaps only thing would be in regards to handling the pin matrix. So, UI/UX wise, perhaps something similar to how shapeshift.com handles it?

![shapeshift-pin](https://user-images.githubusercontent.com/1489224/70825041-8c63f780-1db1-11ea-906e-cd2aa20b59e4.png)

Something for the design team?

**Comment by keepkeyjon** [comment]
`hdwallet-keepkey-chromeusb` is meant for use in a chrome app. `hdwallet-keepkey-webusb` is meant for browsers (requires webusb support, which currently only exists in chrome/brave).

**Comment by keepkeyjon** [comment]
> perhaps only thing would be in regards to handling the pin matrix

Also bip39 passphrases, the input of which gets triggered by a similar event.

**Comment by walidmujahid** [comment]
@Lazaridis I just noticed this comment in my email but it seems to have
been deleted from the thread -and now that I check, your other comments to
the thread as well. It would have been nice to have seen your comments on
technical debt and your link to #7601 when I had checked early this morning.

Also, with your explanation, I understand better why a "fake" keyring route
might be taken.

**Comment by lazaridiscom** [comment]
>> Enough talk, me out here

I delete my messages usually myself after a weak, sometimes faster. Different reasons, here it is mainly that it hurts my brain to discuss stuff which is "obvious grade" for me.

Github leaves no trace if the original author deletes messages (only if the repo-owner deletes them).

So, **I** deleted my messages **myself**, so there is no need for you to reproduce them.

( And now I really really out )

**Comment by keepkeyjon** [comment]
<details><summary>off-topic meta-discussion</summary>

> > > Enough talk, me out here
> 
> I delete my messages usually myself after a weak, sometimes faster. Different reasons, here it is mainly that it hurts my brain to discuss stuff which is "obvious grade" for me.
> 
> Github leaves no trace if the original author deletes messages (only if the repo-owner deletes them).
> 
> So, **I** deleted my messages **myself**, so there is no need for you to reproduce them.
> 
> ( And now I really really out )

@lazaridiscom IMO that revisionist attitude is detrimental to open source projects such as this one. It may be "obvious grade" to you, but not to others... and deleting messages hurts those who come to look at these threads after the fact.

</details>

**Comment by lazaridiscom** [comment]
@keepkeyjon , whatever makes you happy...

**Comment by walidmujahid** [comment]
> The "fake" eth-trezor-keyring could provide this initially by
mock-functionality (even with a hardcoded PIN/Passphrase).

Since reading the previous explanation by @lazaridiscom, I have been
thinking a bit harder on the "fake" keyring and mock-functionality
solution. However, it is the second part, mock-functionality, that I remain
uncertain about.

Was there a method in mind when the idea was proposed as to how to provide
that mock-functionality when comes to the Keepkey pin/passphrase?

I am not sure how a "hardcoded" solution would be implemented and have
people's Keepkeys work with MM. Though, I feel perhaps that "hardcoded"
comment was a solution available to hardware such as Trezor or Ledger? I
have neither, but I do have a Keepkey.

I still feel the other solution path, if you wish to call it so, of
utilising @shapeshiftoss/hdwallet-keepkey{-webusb} packages and not
avoiding initial UI changes seems the most straightforward -even if it may
not the best- when it comes to making things work within a reasonable
amount of time.

Another question, perhaps for @keepkeyjon:
Would I be correct to assume that hdwallet packages could be used within an
implementation of MM's Keyring interface and still use hdwallet's webusb
functionality? Or does MM have to use the hdwallet's keyring in order to
use @shapeshiftoss/hdwallet-keepkey-webusb? Would MM would have to make our
own version of hdwallet-keepkey-webusb and any relevant packages?

**Comment by keepkeyjon** [comment]
> Was there a method in mind when the idea was proposed as to how to provide
that mock-functionality when comes to the Keepkey pin/passphrase?

Mocking passphrase input is possible, but doing so for the PIN is not (because of the cipher pad).

I don't think it makes sense to cut corners the way lazaridiscom was suggesting, though you could use the command-line tools for KeepKey [1] to turn off the PIN for testing.

> Another question, perhaps for @keepkeyjon:
Would I be correct to assume that hdwallet packages could be used within an
implementation of MM's Keyring interface and still use hdwallet's webusb
functionality? Or does MM have to use the hdwallet's keyring in order to
use @shapeshiftoss/hdwallet-keepkey-webusb?

You'll need to use both. To make MM happy, it needs a package that exposes the same interface as eth-trezor-keyring, and that should be implemented on top of hdwallet-keepkey-webbusb + hdwallet-core + hdwallet-keepkey through hdwallet's keyring.

In hdwallet, the keyring is a global singleton for managing all connected wallets (that hdwallet knows about). If you do not register other transports with it, it won't pay attention to the other kinds of devices that hdwallet supports... which is great for this use case, since in MM we don't want hdwallet's keyring to try to manage Trezors/Ledgers, but we do want that where we use it in the ShapeShift platform.

> Would MM would have to make our own version of hdwallet-keepkey-webusb and any relevant packages?

No, don't duplicate the functionality of that... just import it and use it.... that's what it's there for.

1: www.github.com/solipsis/go-keepkey, www.github.com/keepkey/python-keepkey

**Comment by walidmujahid** [comment]
@keepkeyjon Thank you very much for the clarification and links.

**Comment by greatwolf** [comment]
What's the current situation of this request? I would have though KeepKey integration should have happened by now. What are the roadblocks that's stopping this from happening?

**Comment by deepsi43** [comment]
Theres some one else who came up on working on it.Hence didnot check on anything further.
I guess the keyring controller and shapeshift methods are used i order to get the integration done!

**Comment by BitHighlander** [comment]
id like to bump* this. https://github.com/shapeshift/hdwallet should have everything needed to get this done. i'm looking into organizing a sizable bounty for this getting into a release.

**Comment by raj-shekhar1** [comment]
Is this still worked on ?

**Comment by BitHighlander** [comment]
yes, We have identified a limitation with webusb in bex's in general. working on the best way to get around this and landed on a bridge application. After this proposed bridge application is released and popular/tested you can expect a large gitcoin bounty to be posted for this integration work.
